src/device-manager/java/src/nl/Weave/DataManagement/WDMClientImpl.java

```diff
@@ -0,0 +1,213 @@
+/*
+
+    Copyright (c) 2019 Google, LLC.
+    All rights reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+package nl.Weave.DataManagement;
+
+import android.os.Build;
+import android.util.Log;
+
+import java.util.EnumSet;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+
+public class WDMClientImpl implements WDMClient
+{
+    public WDMClientImpl()
+    {
+        init();
+        mCompHandler = null;
+    }
+
+    @Override
+    public void connect(long deviceMgrPtr)
+    {
+        mWDMClientPtr = newWDMClient(deviceMgrPtr);
+    }
+
+    @Override
+    public void close()
+    {
+        if (mTraitMap != null)
+        {
+            Iterator iter = mTraitMap.entrySet().iterator();
+            while (iter.hasNext())
+            {
+                Map.Entry entry = (Map.Entry) iter.next();
+                GenericTraitUpdatableDataSinkImpl traitInstance = (GenericTraitUpdatableDataSinkImpl)entry.getValue();
+                if (traitInstance != null)
+                {
+                    traitInstance.close();
+                }
+            }
+            mTraitMap.clear();
+        }
+
+        if (mWDMClientPtr != 0)
+        {
+            deleteWDMClient(mWDMClientPtr);
+            mWDMClientPtr = 0;
+        }
+
+        mCompHandler = null;
+    }
+
+    @Override
+    public GenericTraitUpdatableDataSinkImpl newDataSink(ResourceIdentifier resourceIdentifier, long profileId, long instanceId, String path)
+    {
+        long traitInstancePtr = 0;
+        boolean isExist = false;
+        GenericTraitUpdatableDataSinkImpl traitInstance = null;
+
+        if (mTraitMap == null)
+        {
+            mTraitMap = new HashMap<Long, GenericTraitUpdatableDataSinkImpl>();
+        }
+
+        if (mWDMClientPtr == 0)
```

Comment:
Can we move the pointer checks to the beginning of the method to fail fast? Otherwise this will leave the map partially initialized.

Reply:
Done, thanks